django/db/backends/postgresql/psycopg_any.py

```diff
@@ -1,31 +1,102 @@
-from enum import IntEnum
+import ipaddress
+from functools import lru_cache
 
-from psycopg2 import errors, extensions, sql  # NOQA
-from psycopg2.extras import DateRange, DateTimeRange, DateTimeTZRange, Inet  # NOQA
-from psycopg2.extras import Json as Jsonb  # NOQA
-from psycopg2.extras import NumericRange, Range  # NOQA
+try:
+    from psycopg import ClientCursor, IsolationLevel, adapt, adapters, errors, sql
+    from psycopg.postgres import types
+    from psycopg.types.datetime import TimestamptzLoader
+    from psycopg.types.json import Jsonb
+    from psycopg.types.range import Range, RangeDumper
+    from psycopg.types.string import TextLoader
 
-RANGE_TYPES = (DateRange, DateTimeRange, DateTimeTZRange, NumericRange)
+    Inet = ipaddress.ip_address
 
+    DateRange = DateTimeRange = DateTimeTZRange = NumericRange = Range
+    RANGE_TYPES = (Range,)
 
-class IsolationLevel(IntEnum):
-    READ_UNCOMMITTED = extensions.ISOLATION_LEVEL_READ_UNCOMMITTED
-    READ_COMMITTED = extensions.ISOLATION_LEVEL_READ_COMMITTED
-    REPEATABLE_READ = extensions.ISOLATION_LEVEL_REPEATABLE_READ
-    SERIALIZABLE = extensions.ISOLATION_LEVEL_SERIALIZABLE
+    TSRANGE_OID = types["tsrange"].oid
+    TSTZRANGE_OID = types["tstzrange"].oid
 
+    def mogrify(sql, params, connection):
+        return ClientCursor(connection.connection).mogrify(sql, params)
 
-def _quote(value, connection=None):
-    adapted = extensions.adapt(value)
-    if hasattr(adapted, "encoding"):
-        adapted.encoding = "utf8"
-    # getquoted() returns a quoted bytestring of the adapted value.
-    return adapted.getquoted().decode()
+    # Adapters.
+    class BaseTzLoader(TimestamptzLoader):
+        """
+        Load a PostgreSQL timestamptz using the a specific timezone.
+        The timezone can be None too, in which case it will be chopped.
+        """
 
+        timezone = None
 
-sql.quote = _quote
+        def load(self, data):
+            res = super().load(data)
+            return res.replace(tzinfo=self.timezone)
```

Comment:
Is this line correct?

I'm doing something weird - using Django's connection params in a non-Django settings (I know it's not recommended, all warranty is off, regular Django not affected). The program's timezone is different from the DB timezone. This line causes the time to be read incorrectly, because it transforms e.g. `2023-04-09 19:44:32.768813+03:00` to `2023-04-09 19:44:32.768813+00:00`, which is a different time.

I wonder if this is what was really meant, or if this wants to be `res.astimezone(tzinfo=self.timezone)` (with special handling of `None`)? That would result in `2023-04-09 16:44:32.768813+00:00`.

Or maybe this assumes the DB timezone == app timezone? But in that case, why is this needed at all?

Reply:
Replacing with `return res` or `return res.astimezone(self.timezone)` causes many tests to fails, e.g. in `queries` or `postgres_tests.test_ranges`. 

> Or maybe this assumes the DB timezone == app timezone? But in that case, why is this needed at all?

As far as I'm aware, this is needed because datetimes are loaded without a timezone so we have to set it manually :thinking:

Reply:
> Replacing with return res or return res.astimezone(self.timezone) causes many tests to fails, e.g. in queries or postgres_tests.test_ranges.

Thanks for checking, I'll try it myself and investigate more.
If you're OK with trying a little tweak, that would save me some effort setting up Django dev env. The docstring says "The timezone can be None too, in which case it will be chopped.", so I think it should be something like this:

```py
res = super().load(data)
if self.timezone is None:
    # USE_TZ=False, convert to system local time zone and make naive.
    return res.astimezone(None).replace(tzinfo=None)
else:
    # USE_TZ=True, convert to configured timezone.
    return res.astimezone(self.timezone)
```

> As far as I'm aware, this is needed because datetimes are loaded without a timezone so we have to set it manually thinking

As far as I can see, psycopg's `TimestamptzLoader` (the super class) always returns an aware datetime in the DB (connection) timezone.

Overall, looking at 4.1.x code, I don't see any messing with timezones other than setting the DB connection's timezone, so I'll try to understand why it's needed in psycopg3 in the first place.

Reply:
> Overall, looking at 4.1.x code, I don't see any messing with timezones other than setting the DB connection's timezone, so I'll try to understand why it's needed in psycopg3 in the first place.

This is probably required in the Django test suite because we're changing the `TIME_ZONE` setting.